File: app/src/main/java/bf/fasobizness/bafatech/activities/annonce/ActivityAnnonceEditPhotos.java
package bf.fasobizness.bafatech.activities.annonce;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;
import com.zhihu.matisse.Matisse;
import com.zhihu.matisse.MimeType;
import com.zhihu.matisse.engine.impl.GlideEngine;
import com.zhihu.matisse.filter.Filter;
import com.zhihu.matisse.internal.entity.CaptureStrategy;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import bf.fasobizness.bafatech.R;
import bf.fasobizness.bafatech.activities.ActivityFullScreen;
import bf.fasobizness.bafatech.adapters.IllustrationEditAdapter;
import bf.fasobizness.bafatech.helper.GifSizeFilter;
import bf.fasobizness.bafatech.helper.ProgressRequestBody;
import bf.fasobizness.bafatech.helper.RetrofitClient;
import bf.fasobizness.bafatech.interfaces.IllustrationInterface;
import bf.fasobizness.bafatech.interfaces.OnItemListener;
import bf.fasobizness.bafatech.interfaces.UploadCallbacks;
import bf.fasobizness.bafatech.models.Announce;
import bf.fasobizness.bafatech.models.MyResponse;
import bf.fasobizness.bafatech.utils.FileCompressingUtil;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.zhihu.matisse.internal.utils.PathUtils.getPath;

public class ActivityAnnonceEditPhotos extends AppCompatActivity implements OnItemListener, UploadCallbacks {
    private static final int REQUEST_CODE = 23;
    private ArrayList<Announce.Annonce.Illustration> images;
    private ArrayList<Uri> uri_images;
    // private UriAdapter mAdapter;
    private String id_annonce = "";
    private IllustrationEditAdapter adapter;
    // private RequestQueue requestQueue;
    private LinearLayout no_picture;
    private ProgressDialog progressDoalog;
    private IllustrationInterface illustrationInterface;

    private static File getFile(Context context, Uri uri) {
        String path = getPath(context, uri);
        if (path != null) {
            if (isLocal(path)) {
                return new File(path);
            }
        }
        return null;
    }

    private static boolean isLocal(String url) {
        return url != null && !url.startsWith("http://") && !url.startsWith("https://");
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_annonce_edit_photos);

        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(getString(R.string.modification));
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        toolbar.setNavigationIcon(R.drawable.left_white);
        toolbar.setNavigationOnClickListener(view -> finish());

        no_picture = findViewById(R.id.layout_no_picture);
        no_picture.setOnClickListener(v -> requestMultiplePermissions());

        illustrationInterface = RetrofitClient.getClient().create(IllustrationInterface.class);

        images = new ArrayList<>();
        uri_images = new ArrayList<>();
        RecyclerView recyclerView = findViewById(R.id.recyclerview_photos);
        adapter = new IllustrationEditAdapter(this, images);
        LinearLayoutManager manager = new GridLayoutManager(getApplicationContext(), 2);
        recyclerView.setAdapter(adapter);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(manager);
        adapter.setOnItemListener(this);

        Intent extras = getIntent();
        Announce.Annonce annonce = (Announce.Annonce) extras.getSerializableExtra("annonce");


        List<Announce.Annonce.Illustration> arrayList;
        if (annonce != null) {
            arrayList = annonce.illustrations;
            images.addAll(arrayList);
            adapter.notifyDataSetChanged();
            id_annonce = annonce.getId_ann();
        }
        countPictures();

    }

    private void countPictures() {
        if (images.size() == 0) {
            no_picture.setVisibility(View.VISIBLE);
        } else {
            no_picture.setVisibility(View.GONE);
        }
    }

    @Override
    public void onItemClicked(int position) {
        ArrayList<String> imagesList = new ArrayList<>();
        for (int i = 0; i < images.size(); i++) {
            Announce.Annonce.Illustration illustration = images.get(i);
            imagesList.add(illustration.getNom());
        }
        Intent intent = new Intent(this, ActivityFullScreen.class);
        intent.putStringArrayListExtra("images", imagesList);
        intent.putExtra("position", position);
        startActivity(intent);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_illustration_edit, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.nav_illustration_add) {
            requestMultiplePermissions();
        } else if (id == R.id.nav_illustration_remove) {
            ArrayList<Announce.Annonce.Illustration> checked = adapter.getChecked();
            StringBuilder idList = new StringBuilder();
            for (int i = 0; i < checked.size(); i++) {
                idList.append(checked.get(i).getId_illustration()).append(",");
                images.remove(checked.get(i));
                adapter.notifyDataSetChanged();
            }
            delete(idList.toString());
            checked.clear();
            adapter.notifyDataSetChanged();
            countPictures();
        }

        return super.onOptionsItemSelected(item);
    }

    private void delete(String ids) {
        Log.d("Activity", ids);
        Call<MyResponse> call = illustrationInterface.deleteIllustrations(ids);
        call.enqueue(new Callback<MyResponse>() {
            @Override
            public void onResponse(@NonNull Call<MyResponse> call, @NonNull Response<MyResponse> response) {
                Log.d("Activity", response.toString());
                Toast.makeText(ActivityAnnonceEditPhotos.this, "Super", Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onFailure(@NonNull Call<MyResponse> call, @NonNull Throwable t) {
                Log.d("Activty", t.toString());
            }
        });
    }

    private void showChooser() {
        Matisse.from(ActivityAnnonceEditPhotos.this)
                .choose(MimeType.ofAll(), false)
                // .choose(MimeType.ofImage(), false)
                .theme(R.style.Matisse_Dracula)
                .countable(true)
                .capture(true)
                .captureStrategy(
                        new CaptureStrategy(true, "bf.fasobizness.bafatech.fileprovider", "Pictures"))
                .maxSelectable(9)
                .addFilter(new GifSizeFilter(320, 320, 5 * Filter.K * Filter.K))
                .gridExpectedSize(
                        getResources().getDimensionPixelSize(R.dimen.grid_expected_size))
                .restrictOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT)
                .thumbnailScale(0.85f)
                .imageEngine(new GlideEngine())
                .showSingleMediaType(true)
                .originalEnable(true)
                .maxOriginalSize(10)
                .autoHideToolbarOnSingleTap(true)
                .forResult(REQUEST_CODE);
    }

    private void requestMultiplePermissions() {
        Dexter.withActivity(this)
                .withPermissions(
                        Manifest.permission.CAMERA,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.READ_EXTERNAL_STORAGE)
                .withListener(new MultiplePermissionsListener() {
                    @Override
                    public void onPermissionsChecked(MultiplePermissionsReport report) {
                        if (report.areAllPermissionsGranted()) {
                            showChooser();
                        }
                    }

                    @Override
                    public void onPermissionRationaleShouldBeShown(List<PermissionRequest> permissions, PermissionToken token) {
                        token.continuePermissionRequest();
                    }
                }).
                withErrorListener(error -> Toast.makeText(getApplicationContext(), "Some Error! ", Toast.LENGTH_SHORT).show())
                .onSameThread()
                .check();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE && resultCode == RESULT_OK) {
            if (data != null) {
                uri_images.addAll(Matisse.obtainResult(data));

                uploadPictures();
            }
        }
    }

    @Override
    public void onProgressUpdate(int percentage) {
        // Toast.makeText(this, percentage+"%", Toast.LENGTH_SHORT).show();
        progressDoalog.setMessage(percentage + "%");
    }

    @Override
    public void onError() {
    }

    @Override
    public void onFinish() {
        progressDoalog.dismiss();
    }

    private void getAllPictures() {
        images.clear();
        adapter.notifyDataSetChanged();

        if (!id_annonce.isEmpty()) {
            IllustrationInterface illustrationInterface = RetrofitClient.getClient().create(IllustrationInterface.class);
            Call<Announce.Annonce> call = illustrationInterface.getIllustrations(id_annonce);
            call.enqueue(new Callback<Announce.Annonce>() {
                @Override
                public void onResponse(@NonNull Call<Announce.Annonce> call, @NonNull Response<Announce.Annonce> response) {
                    if (response.isSuccessful()) {

                        Announce.Annonce annonce = response.body();
                        List<Announce.Annonce.Illustration> arrayList = null;
                        if (annonce != null) {
                            arrayList = annonce.illustrations;
                        }
                        if (arrayList != null) {
                            images.addAll(arrayList);
                        }

                    } else {
                        Toast.makeText(ActivityAnnonceEditPhotos.this, R.string.une_erreur_sest_produite, Toast.LENGTH_SHORT).show();
                    }

                    countPictures();
                    adapter.notifyDataSetChanged();
                }

                @Override
                public void onFailure(@NonNull Call<Announce.Annonce> call, @NonNull Throwable t) {
                    Toast.makeText(ActivityAnnonceEditPhotos.this, R.string.pas_d_acces_internet, Toast.LENGTH_SHORT).show();
                }
            });
        }
        /*String url = Constants.HOST_URL + "illustration/read/" + id_annonce;
        StringRequest request = new StringRequest(Request.Method.GET, url, response -> {
            try {
                images.clear();
                JSONObject jsonObject = new JSONObject(response);
                JSONArray jsonArray = jsonObject.getJSONArray("data");
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject data = jsonArray.getJSONObject(i);
                    Announce.Annonce.Illustration illustration = new Illustration();
                    illustration.setId(data.getString("id_illustration"));
                    illustration.setNom(data.getString("nom"));
                    images.add(illustration);
                }
                adapter.notifyDataSetChanged();
                countPictures();
            } catch (JSONException e) {
                e.printStackTrace();
                Toast.makeText(this, R.string.une_erreur_sest_produite, Toast.LENGTH_SHORT).show();
            }

        }, error -> Toast.makeText(this, R.string.pas_d_acces_internet, Toast.LENGTH_SHORT).show());
        requestQueue.add(request);*/
    }

    @Override
    public void uploadStart() {
    }

    private MultipartBody.Part prepareFilePart(String part, Uri uri) {
        File file = getFile(this, uri);
        assert file != null;
        FileCompressingUtil fileCompressingUtil = new FileCompressingUtil();
        File compressedFile = fileCompressingUtil.saveBitmapToFile(file);
        ProgressRequestBody requestFile = new ProgressRequestBody(compressedFile, this);
        return MultipartBody.Part.createFormData(part, file.getName(), requestFile);
    }

    private RequestBody createPart(String id_ann) {
        return RequestBody.create(
                okhttp3.MultipartBody.FORM, id_ann);
    }

    private void uploadPictures() {
        progressDoalog = new ProgressDialog(this);
        progressDoalog.setTitle(R.string.chargement_en_cours);
        progressDoalog.show();

        List<MultipartBody.Part> parts = new ArrayList<>();
        for (int i = 0; i < uri_images.size(); i++) {
            parts.add(prepareFilePart("image" + i, uri_images.get(i)));
        }
        RequestBody id_ann = createPart(id_annonce);
        RequestBody size = createPart(parts.size() + "");
        Call<ResponseBody> call = illustrationInterface.uploadPhotos(id_ann, size, parts);
        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(@NonNull Call<ResponseBody> call, @NonNull Response<ResponseBody> response) {
                Log.d("ActivityEdit", response.toString());
                if (response.isSuccessful()) {
                    progressDoalog.dismiss();
                    Toast.makeText(ActivityAnnonceEditPhotos.this, R.string.succes, Toast.LENGTH_SHORT).show();
                    getAllPictures();
                } else {
                    Toast.makeText(ActivityAnnonceEditPhotos.this, R.string.une_erreur_sest_produite, Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable t) {
                progressDoalog.dismiss();
                Toast.makeText(ActivityAnnonceEditPhotos.this, R.string.pas_d_acces_internet, Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/bf/fasobizness/bafatech/activities/annonce/ActivityAnnonceEditPhotos.kt
package bf.fasobizness.bafatech.activities.annonce

import android.Manifest
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.content.pm.ActivityInfo
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.LinearLayout
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import bf.fasobizness.bafatech.R
import bf.fasobizness.bafatech.activities.ActivityFullScreen
import bf.fasobizness.bafatech.adapters.IllustrationEditAdapter
import bf.fasobizness.bafatech.helper.GifSizeFilter
import bf.fasobizness.bafatech.helper.ProgressRequestBody
import bf.fasobizness.bafatech.helper.RetrofitClient
import bf.fasobizness.bafatech.interfaces.IllustrationInterface
import bf.fasobizness.bafatech.interfaces.OnItemListener
import bf.fasobizness.bafatech.interfaces.UploadCallbacks
import bf.fasobizness.bafatech.models.Announce.Annonce
import bf.fasobizness.bafatech.models.Announce.Annonce.Illustration
import bf.fasobizness.bafatech.models.MyResponse
import com.karumi.dexter.Dexter
import com.karumi.dexter.MultiplePermissionsReport
import com.karumi.dexter.PermissionToken
import com.karumi.dexter.listener.DexterError
import com.karumi.dexter.listener.PermissionRequest
import com.karumi.dexter.listener.multi.MultiplePermissionsListener
import com.nguyenhoanglam.imagepicker.model.Config
import com.nguyenhoanglam.imagepicker.model.Image
import com.nguyenhoanglam.imagepicker.ui.imagepicker.ImagePicker
import com.zhihu.matisse.Matisse
import com.zhihu.matisse.MimeType
import com.zhihu.matisse.engine.impl.GlideEngine
import com.zhihu.matisse.filter.Filter
import com.zhihu.matisse.internal.entity.CaptureStrategy
import com.zhihu.matisse.internal.utils.PathUtils
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.File
import java.util.*

class ActivityAnnonceEditPhotos : AppCompatActivity(), OnItemListener, UploadCallbacks {
    private lateinit var images: ArrayList<Illustration>

    //private lateinit var uri_images: ArrayList<Uri>
    private var localImages: ArrayList<Image> = ArrayList()

    // private UriAdapter mAdapter;
    private var id_annonce = ""
    private lateinit var adapter: IllustrationEditAdapter

    // private RequestQueue requestQueue;
    private lateinit var no_picture: LinearLayout
    private lateinit var progressDoalog: ProgressDialog
    private lateinit var illustrationInterface: IllustrationInterface
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_annonce_edit_photos)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.title = getString(R.string.modification)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        toolbar.setNavigationIcon(R.drawable.left_white)
        toolbar.setNavigationOnClickListener { finish() }
        no_picture = findViewById(R.id.layout_no_picture)
        no_picture.setOnClickListener { requestMultiplePermissions() }
        illustrationInterface = RetrofitClient.getClient().create(IllustrationInterface::class.java)
        images = ArrayList()
        localImages = ArrayList()
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerview_photos)
        adapter = IllustrationEditAdapter(this, images)
        val manager: LinearLayoutManager = GridLayoutManager(applicationContext, 2)
        recyclerView.adapter = adapter
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = manager
        adapter.setOnItemListener(this)
        val extras = intent
        val annonce = extras.getSerializableExtra("annonce") as Annonce?
        val arrayList: List<Illustration>
        if (annonce != null) {
            arrayList = annonce.illustrations
            images.addAll(arrayList)
            adapter.notifyDataSetChanged()
            id_annonce = annonce.id_ann
        }
        countPictures()
    }

    private fun countPictures() {
        if (images.size == 0) {
            no_picture.visibility = View.VISIBLE
        } else {
            no_picture.visibility = View.GONE
        }
    }

    override fun onItemClicked(position: Int) {
        val imagesList = ArrayList<String>()
        for (i in images.indices) {
            val illustration = images[i]
            imagesList.add(illustration.nom)
        }
        val intent = Intent(this, ActivityFullScreen::class.java)
        intent.putStringArrayListExtra("images", imagesList)
        intent.putExtra("position", position)
        startActivity(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_illustration_edit, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.nav_illustration_add) {
            requestMultiplePermissions()
        } else if (id == R.id.nav_illustration_remove) {
            val checked = adapter.checked
            val idList = StringBuilder()
            for (i in checked.indices) {
                idList.append(checked[i].id_illustration).append(",")
                images.remove(checked[i])
                adapter.notifyDataSetChanged()
            }
            delete(idList.toString())
            checked.clear()
            adapter.notifyDataSetChanged()
            countPictures()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun delete(ids: String) {
        Log.d("Activity", ids)
        val call = illustrationInterface.deleteIllustrations(ids)
        call.enqueue(object : Callback<MyResponse?> {
            override fun onResponse(call: Call<MyResponse?>, response: Response<MyResponse?>) {
                Log.d("Activity", response.toString())
                Toast.makeText(this@ActivityAnnonceEditPhotos, "Super", Toast.LENGTH_SHORT).show()
            }

            override fun onFailure(call: Call<MyResponse?>, t: Throwable) {
                Log.d("Activty", t.toString())
            }
        })
    }

    private fun showChooser() {
        ImagePicker.with(this)
                .setFolderMode(true)
                .setFolderTitle("Photos")
                .setRootDirectoryName(Config.ROOT_DIR_DCIM)
                .setDirectoryName("Faso Biz Ness")
                .setMultipleMode(true)
                .setShowNumberIndicator(true)
                .setMaxSize(10)
                .setLimitMessage("Vous pouvez selectionner 10 images")
                .setSelectedImages(localImages)
                .setRequestCode(100)
                .setShowCamera(false)
                .start()
        /*Matisse.from(this@ActivityAnnonceEditPhotos)
                .choose(MimeType.ofAll(), false) // .choose(MimeType.ofImage(), false)
                .theme(R.style.Matisse_Dracula)
                .countable(true)
                .capture(false)
                .captureStrategy(
                        CaptureStrategy(true, "bf.fasobizness.bafatech.fileprovider", "Pictures"))
                .maxSelectable(9)
                .addFilter(GifSizeFilter(320, 320, 5 * Filter.K * Filter.K))
                .gridExpectedSize(
                        resources.getDimensionPixelSize(R.dimen.grid_expected_size))
                .restrictOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT)
                .thumbnailScale(0.85f)
                .imageEngine(GlideEngine())
                .showSingleMediaType(true)
                .originalEnable(true)
                .maxOriginalSize(10)
                .autoHideToolbarOnSingleTap(true)
                .forResult(REQUEST_CODE)*/
    }

    private fun requestMultiplePermissions() {
        Dexter.withContext(this)
                .withPermissions(
                        Manifest.permission.CAMERA,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        Manifest.permission.READ_EXTERNAL_STORAGE)
                .withListener(object : MultiplePermissionsListener {
                    override fun onPermissionsChecked(report: MultiplePermissionsReport) {
                        if (report.areAllPermissionsGranted()) {
                            showChooser()
                        }
                    }

                    override fun onPermissionRationaleShouldBeShown(permissions: List<PermissionRequest>, token: PermissionToken) {
                        token.continuePermissionRequest()
                    }
                }).withErrorListener { Toast.makeText(applicationContext, "Some Error! ", Toast.LENGTH_SHORT).show() }
                .onSameThread()
                .check()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (ImagePicker.shouldHandleResult(requestCode, resultCode, data, 100)) {
            images.clear()
            adapter.clearAll()
            adapter.notifyDataSetChanged()
            localImages.addAll(ImagePicker.getImages(data))

            images.clear()
            adapter.clearAll()
            // uploadPictures()
            // adapter.notifyDataSetChanged()
        }
        /*if (requestCode == REQUEST_CODE && resultCode == RESULT_OK) {
            if (data != null) {
                uri_images.addAll(Matisse.obtainResult(data))
                uploadPictures()
            }
        }*/
    }

    override fun onProgressUpdate(percentage: Int) {
        // Toast.makeText(this, percentage+"%", Toast.LENGTH_SHORT).show();
        progressDoalog.setMessage("$percentage%")
    }

    override fun onError() {}
    override fun onFinish() {
        progressDoalog.dismiss()
    }

    private fun allPictures() {
        localImages.clear()
        images.clear()
        adapter.clearAll()
        adapter.notifyDataSetChanged()
        if (id_annonce.isNotEmpty()) {
            val illustrationInterface = RetrofitClient.getClient().create(IllustrationInterface::class.java)
            val call = illustrationInterface.getIllustrations(id_annonce)
            call.enqueue(object : Callback<Annonce?> {
                override fun onResponse(call: Call<Annonce?>, response: Response<Annonce?>) {
                    if (response.isSuccessful) {
                        val annonce = response.body()
                        var arrayList: List<Illustration>? = null
                        if (annonce != null) {
                            arrayList = annonce.illustrations
                        }
                        if (arrayList != null) {
                            images.addAll(arrayList)
                        }
                    } else {
                        Toast.makeText(this@ActivityAnnonceEditPhotos, R.string.une_erreur_sest_produite, Toast.LENGTH_SHORT).show()
                    }
                    countPictures()
                    adapter!!.notifyDataSetChanged()
                }

                override fun onFailure(call: Call<Annonce?>, t: Throwable) {
                    Toast.makeText(this@ActivityAnnonceEditPhotos, R.string.pas_d_acces_internet, Toast.LENGTH_SHORT).show()
                }
            })
        }
    }

    override fun uploadStart() {}
    private fun prepareFilePart(part: String, uri: Uri): MultipartBody.Part {
        val file = getFile(this, uri)!!
        // FileCompressingUtil fileCompressingUtil = new FileCompressingUtil();
        // File compressedFile = fileCompressingUtil.saveBitmapToFile(file);
        val requestFile = ProgressRequestBody(file, this)
        return MultipartBody.Part.createFormData(part, file.name, requestFile)
    }

    private fun createPart(id_ann: String): RequestBody {
        return RequestBody.create(
                MultipartBody.FORM, id_ann)
    }

    private fun uploadPictures() {
        progressDoalog = ProgressDialog(this)
        progressDoalog.setTitle(R.string.chargement_en_cours)
        progressDoalog.show()
        localImages.clear()
        images.clear()
        adapter.clearAll()
        adapter.notifyDataSetChanged()
        val parts: MutableList<MultipartBody.Part> = ArrayList()
        for (i in localImages.indices) {
            parts.add(prepareFilePart("image$i", localImages[i].uri))
        }
        val id_ann = createPart(id_annonce)
        val size = createPart(parts.size.toString() + "")
        val call = illustrationInterface.uploadPhotos(id_ann, size, parts)
        call.enqueue(object : Callback<ResponseBody?> {
            override fun onResponse(call: Call<ResponseBody?>, response: Response<ResponseBody?>) {
                Log.d("ActivityEdit", response.toString())
                if (response.isSuccessful) {
                    progressDoalog.dismiss()
                    allPictures()
                } else {
                    Toast.makeText(this@ActivityAnnonceEditPhotos, R.string.une_erreur_sest_produite, Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<ResponseBody?>, t: Throwable) {
                progressDoalog.dismiss()
                Toast.makeText(this@ActivityAnnonceEditPhotos, R.string.pas_d_acces_internet, Toast.LENGTH_SHORT).show()
            }
        })
    }

    companion object {
        private const val REQUEST_CODE = 23
        private fun getFile(context: Context, uri: Uri): File? {
            val path = PathUtils.getPath(context, uri)
            if (path != null) {
                if (isLocal(path)) {
                    return File(path)
                }
            }
            return null
        }

        private fun isLocal(url: String?): Boolean {
            return url != null && !url.startsWith("http://") && !url.startsWith("https://")
        }
    }
}